JavaScript 中哪一种循环更快呢？ (#8048)
xitu/gold-miner: article/2021/which-type-of-loop-is-fastest-in-javascript.md

> * 原文地址：[Which type of loop is fastest in JavaScript?](https://medium.com/javascript-in-plain-english/which-type-of-loop-is-fastest-in-javascript-ec834a0f21b9)
> * 原文作者：[kushsavani](https://kushsavani.medium.com/)
> * 译文出自：[掘金翻译计划](https://github.com/xitu/gold-miner)
> * 本文永久链接：[https://github.com/xitu/gold-miner/blob/master/article/2021/which-type-of-loop-is-fastest-in-javascript.md](https://github.com/xitu/gold-miner/blob/master/article/2021/which-type-of-loop-is-fastest-in-javascript.md)
> * 译者：[霜羽 Hoarfroster](https://github.com/PassionPenguin)
> * 校对者：

# JavaScript 中哪一种循环更快呢？

了解哪一种 `for` 循环或迭代器适合我们的需求，防止我们犯下一些低级错误，让我们的应用程序性能受到影响。

> [由 [Artem Sapegin](https://unsplash.com/@sapegin?utm_source=medium&utm_medium=referral) 上传至 [Unsplash](https://unsplash.com/?utm_source=medium&utm_medium=referral)](https://miro.medium.com/max/10944/0*FjGuCxH-seN1PrRF)

Author: 看起来是的（

> JavaScript 是 Web 开发的一种新的重磅引发者。JavaScript 不仅拥有着巨额的框架（如 Node.js、React、Angular 又或是 Vue 等），而且 Vanilla JavaScript 也拥有庞大的支持者。让我们一起走进 JavaScript。我们都清楚，循环一直是大多数编程语言的重要组成部分，而 JavaScript 为我们提供了许多迭代或循环值的方法。

Author: new sensation 一开始我的想法是，搞大事情的那个东东，于是就...

> JavaScript 是 Web 开发的一种新的重磅引发者。JavaScript 不仅拥有着巨额的框架（如 Node.js、React、Angular 又或是 Vue 等），而且 Vanilla JavaScript 也拥有庞大的支持者。让我们一起走进 JavaScript。我们都清楚，循环一直是大多数编程语言的重要组成部分，而 JavaScript 为我们提供了许多迭代或循环值的方法。

Author: large fan base 翻译为 大量的框架的确不应该（，打脸打脸

> 但是问题是，我们是否真的知道哪个循环或迭代最适合我们的需求。`for` 循环有很多变形，例如 `for`、`for`（倒序）、`for` `of`、`forEach`、`for` `in`、`for` `await`。本文将涵盖此类辩论。

Author: 第一句话个人没感觉任何问题诶，看到的出版物中也有挺多的

> 但是问题是，我们是否真的知道哪个循环或迭代最适合我们的需求。`for` 循环有很多变形，例如 `for`、`for`（倒序）、`for` `of`、`forEach`、`for` `in`、`for` `await`。本文将涵盖此类辩论。

Reviewer: > 第一句话个人没感觉任何问题诶，看到的出版物中也有挺多的

这里的重点是读起来是否流畅，你可以读几遍看看，”但问题在于“是更加顺口的。

> 但是问题是，我们是否真的知道哪个循环或迭代最适合我们的需求。`for` 循环有很多变形，例如 `for`、`for`（倒序）、`for` `of`、`forEach`、`for` `in`、`for` `await`。本文将涵盖此类辩论。

Author: > > 第一句话个人没感觉任何问题诶，看到的出版物中也有挺多的
> 
> 这里的重点是读起来是否流畅，你可以读几遍看看，”但问题在于“是更加顺口的。

没有问题的，我们平时都这样说话的，至少在附近环境中

> 但是问题是，我们是否真的知道哪个循环或迭代最适合我们的需求。`for` 循环有很多变形，例如 `for`、`for`（倒序）、`for` `of`、`forEach`、`for` `in`、`for` `await`。本文将涵盖此类辩论。

Reviewer 2: 连着出现几个相同的字读出来会有绕口的感觉，也会对阅读产生`韵律因素`上的影响。这个是文学上的要求了，可以参考着改（哈哈）

> 但是问题是，我们是否真的知道哪个循环或迭代最适合我们的需求。`for` 循环有很多变形，例如 `for`、`for`（倒序）、`for` `of`、`forEach`、`for` `in`、`for` `await`。本文将涵盖此类辩论。

Author: > 连着出现几个相同的字读出来会有绕口的感觉，也会对阅读产生`韵律因素`上的影响。这个是文学上的要求了，可以参考着改（哈哈）

\(//∇//)\ 这一句话也才两个`是`啊 ( ; _ ; )/

> 但是问题是，我们是否真的知道哪个循环或迭代最适合我们的需求。`for` 循环有很多变形，例如 `for`、`for`（倒序）、`for` `of`、`forEach`、`for` `in`、`for` `await`。本文将涵盖此类辩论。

Reviewer 2: > (//∇//)\ 这一句话也才两个是啊 ( ; _ ; )/

这里是可改可不改的那种建议（个人倾向于校对的写法，改了也不会有什么损失...）

> 但是问题是，我们是否真的知道哪个循环或迭代最适合我们的需求。`for` 循环有很多变形，例如 `for`、`for`（倒序）、`for` `of`、`forEach`、`for` `in`、`for` `await`。本文将涵盖此类辩论。

Author: 反正最后也改掉了（

## **究竟哪一种循环更快？**

**答案其实是：** `for`（倒序）

这才是令人惊讶的事情！我在我的计算机上测试了一下它们，而我最后发现，也被迫接受这个现实，`for`（倒序）是所有 `for` 循环中最快的。让我分享一个例子：对一个包含超过一百万项元素的 `array` 执行循环。

**免责声明**：结果的准确性在很大程度上取决于我们运行测试的系统配置。在[此处](https://johnresig.com/blog/accuracy-of-javascript-time/)仔细了解准确性。

```
const million = 1000000; 
const arr = Array(million);
console.time('⏳');
for (let i = arr.length; i > 0; i--) {} // for(倒序) 	:- 1.5ms
for (let i = 0; i < arr.length; i++) {} // for          :- 1.6ms
arr.forEach(v => v)                     // foreach      :- 2.1ms
for (const v of arr) {}                 // for...of     :- 11.7ms
console.timeEnd('⏳');
```

> 造成这样结果的原因很简单，在代码中，正序或倒序的 `for` 循环几乎花费一样的时间，仅仅相差了 0.1 毫秒。原因是，`for`（倒序）只需要计算一次起始变量 `let i = arr.length`，而在正序的 `for` 循环中，它在每次变量增加后都会检查条件 `i<arr.length`。对于小量数据来说，我们大可忽略它，但是根据译者的测试，当数据量扩大，例如十亿，千亿等的数量级，差距就显著提升。 而 `forEach` 是 `Array` 原型的一种方法。与普通的 `for` 循环相比，`forEach` 和 `for` `of` 需要花费更多的时间进行数组迭代，但值得注意的是，`for` `of` 和 `forEach` 都获取了数据，而原型并没有，因此没有可比性。

Author: 这里是原作的失误，理应补充，因为 for…of  for…in和 forEach 都不仅仅是循环而且还从对象中提取了数据，而 for 的版本中原作者代码没有任何获取数据的部分

```suggestion
造成这样结果的原因很简单，在代码中，正序和倒序的 `for` 循环几乎花费一样的时间，仅仅相差了 0.1 毫秒。原因是，`for`（倒序）只需要计算一次起始变量 `let i = arr.length`，而在正序的 `for` 循环中，它在每次变量增加后都会检查条件 `i<arr.length`。这个细微的差别不是很重要，你可以忽略它。（译者注：在数据量小或对时间不敏感的代码上，我们大可忽略它，但是根据译者的测试，当数据量扩大，例如十亿，千亿等的数量级，差距就显著提升，我们就需要考虑时间对应用程序性能的影响了。）

而 `forEach` 是 `Array` 原型的一个方法，与普通的 `for` 循环相比，`forEach` 和 `for…of` 需要花费更多的时间进行数组迭代。（译者注：但值得注意的是，`for…of` 和 `forEach` 都从对象中获取了数据，而原型并没有，因此没有可比性。）
```

## **循环的类型，以及我们应该在何处使用它们**

### **1. For loop (forward and reverse)**

> 我想，所有人都应该对这个基础循环非常熟悉了。我们可以在任何我们需要的地方使用 `for` 循环，按照核定的次数运行一段代码。最基础的 `for` 循环运行最迅速的，那么我们应该就这样使用？并不然，性能不仅仅只是唯一标准，我们还需要注意代码可读性，就让我们按照符合我们应用程序的版本就行了。

Author: 并不然，算是通用语，然 表 是这样

```suggestion
我想，每个人都应该对这个循环非常熟悉了。我们可以在任何需要的地方使用 `for` 循环，按照指定的次数重复运行一段代码。最基础的 `for` 循环运行速度往往是最快的，那么我们就应该在任何地方都使用它吗？并不是这样的，因为性能并非总是唯一的标准，**代码可读性**也是很重要的，所以我们要选择适合自己应用的代码。
```

另外代码可读性可以通过注释解决，另外for的代码可读性也很好，要分清楚情况，这里的 usually 应该改为在特定情况下，完全不同意原作的观点。

> 我想，所有人都应该对这个基础循环非常熟悉了。我们可以在任何我们需要的地方使用 `for` 循环，按照核定的次数运行一段代码。最基础的 `for` 循环运行最迅速的，那么我们应该就这样使用？并不然，性能不仅仅只是唯一标准，我们还需要注意代码可读性，就让我们按照符合我们应用程序的版本就行了。

Reviewer: > 并不然，算是通用语，然 表 是这样
> 
> 另外代码可读性可以通过注释解决，另外for的代码可读性也很好，要分清楚情况，这里的 usually 应该改为在特定情况下，完全不同意原作的观点。

1. 这是白话文不是文言文，”并不然“的翻译让人一头雾水
2. 注释只是做适当的补充说明，它的本意并不在于提高代码的可读性。而且如果代码真的很烂的话，我不认为注释是万能药。我也不同意作者的观点，并且我认为”代码可读性也是很重要的“是更符合事实的翻译，但我还是之前的观点，如果不认可他的说法，那么就在括弧里表明这一点，而不是直接修改译文。译文是需要传达作者的原意的，即便作者的观点是错误或者片面的，也要传达给读者，只是你需要告诉读者这个观点可能是错误的。

> 我想，所有人都应该对这个基础循环非常熟悉了。我们可以在任何我们需要的地方使用 `for` 循环，按照核定的次数运行一段代码。最基础的 `for` 循环运行最迅速的，那么我们应该就这样使用？并不然，性能不仅仅只是唯一标准，我们还需要注意代码可读性，就让我们按照符合我们应用程序的版本就行了。

Author: 并不然的话，你可以搜一下百度，结果也不少的，古今都是一个意义，平时都这样说，输入法也有这样的拼写（Apple、GBoard）
不然这个词也有解释，不是这样并非如此的意思，可以参考有道词典

> 我想，所有人都应该对这个基础循环非常熟悉了。我们可以在任何我们需要的地方使用 `for` 循环，按照核定的次数运行一段代码。最基础的 `for` 循环运行最迅速的，那么我们应该就这样使用？并不然，性能不仅仅只是唯一标准，我们还需要注意代码可读性，就让我们按照符合我们应用程序的版本就行了。

Reviewer: > 并不然的话，你可以搜一下百度，结果也不少的，古今都是一个意义，平时都这样说，输入法也有这样的拼写（Apple、GBoard）
> 不然这个词也有解释，不是这样并非如此的意思，可以参考有道词典

平时说的是“其实并不然”，而不是单独的一个“并不然”。对于“不然”这个词，我的理解是和“非也”、“并非如此”一样的意思，但单独拿“并不然”出来使用，恐怕不符合我们的日常用法。或者你如果方便的话，可以提供几个单独使用“并不然”的例子。

> 我想，所有人都应该对这个基础循环非常熟悉了。我们可以在任何我们需要的地方使用 `for` 循环，按照核定的次数运行一段代码。最基础的 `for` 循环运行最迅速的，那么我们应该就这样使用？并不然，性能不仅仅只是唯一标准，我们还需要注意代码可读性，就让我们按照符合我们应用程序的版本就行了。

Author: 对啊【非也，不是这样，并非如此】这样的意思啊，没毛病啊

> 我想，所有人都应该对这个基础循环非常熟悉了。我们可以在任何我们需要的地方使用 `for` 循环，按照核定的次数运行一段代码。最基础的 `for` 循环运行最迅速的，那么我们应该就这样使用？并不然，性能不仅仅只是唯一标准，我们还需要注意代码可读性，就让我们按照符合我们应用程序的版本就行了。

Reviewer: > 对啊【非也，不是这样，并非如此】这样的意思啊，没毛病啊

我说了这不符合日常使用习惯，你在哪里见过**单独使用**并不然的用法吗？

### **2. `forEach`**

> 这个函数需要提供一个回调函数，回调函数会对每一个元素以元素和元素索引（`T`, `Int`）两个值作为参数赋予给我们提供的回调函数并执行这个函数，其中元素索引是可选的参数。

Author: 1. 方法也是函数，函数包含方法，这里的确扩大范围了
2. 这可不是个病句，【回调函数】（对）元素..作参数（赋予）【回调函数】并（执行）【函数】

```suggestion
这个方法需要接受一个回调函数作为输入参数，遍历数组的每一个元素，执行我们的回调函数（以元素本身和它的索引（可选参数）作为参数赋予给回调函数）。`forEach` 还可接受一个可选参数 `this` 供回调函数使用。  
```

> 这个函数需要提供一个回调函数，回调函数会对每一个元素以元素和元素索引（`T`, `Int`）两个值作为参数赋予给我们提供的回调函数并执行这个函数，其中元素索引是可选的参数。

Reviewer: > 1. 方法也是函数，函数包含方法，这里的确扩大范围了
> 2. 这可不是个病句，【回调函数】（对）元素..作参数（赋予）【回调函数】并（执行）【函数】

”回调函数将xxx作为参数赋予给回调函数并执行这个函数。”很明显的句式杂糅，不难看出是病句吧？而且不是回调函数把参数赋予给自己，而是我们把参数传给回调函数，回调函数再负责去接收而已，

> 这个函数需要提供一个回调函数，回调函数会对每一个元素以元素和元素索引（`T`, `Int`）两个值作为参数赋予给我们提供的回调函数并执行这个函数，其中元素索引是可选的参数。

Author: 杂糅了啥，怎么拆开？拆不了嗷

回调函数不能赋予给自己，是方法赋予给回调函数参数呀

> 这个函数需要提供一个回调函数，回调函数会对每一个元素以元素和元素索引（`T`, `Int`）两个值作为参数赋予给我们提供的回调函数并执行这个函数，其中元素索引是可选的参数。

Reviewer: > 杂糅了啥，怎么拆开？拆不了嗷
> 
> 回调函数不能赋予给自己，是方法赋予给回调函数参数呀

一句话出现三个“回调函数”，不是句式杂糅是什么呢。？为什么拆开不了，难道一定要在一句话里把整个意思表达清楚吗？适当灵活变通。
你自己都说了是方法给回调函数参数，那请仔细阅读这句话“回调函数会对每一个元素以元素和元素索引（`T`, `Int`）两个值作为参数赋予给我们提供的回调函数并执行这个函数”，你这句话的意思是回调函数给了回调函数参数。

> 这个函数需要提供一个回调函数，回调函数会对每一个元素以元素和元素索引（`T`, `Int`）两个值作为参数赋予给我们提供的回调函数并执行这个函数，其中元素索引是可选的参数。

Author: 诶嘿好像真的是（

> 这个函数需要提供一个回调函数，回调函数会对每一个元素以元素和元素索引（`T`, `Int`）两个值作为参数赋予给我们提供的回调函数并执行这个函数，其中元素索引是可选的参数。

Author: 完了我乱了抱歉抱歉

```
const things = ['have', 'fun', 'coding'];
const callbackFun = (item, idex) => {
    console.log(`${item} - ${index}`);
}
things.foreach(callbackFun); 
/* 输出 	have - 0
      	fun - 1
      	coding - 2 */
```

需要注意的是，如果我们要使用 `forEach`，我们不能使用 JavaScript 的短路运算符，即不能在每一次循环中跳过或结束循环。

### **3. `for` `of`**

`for` `of` 作为 ECMAScript 6 的标准内容，会对一个可迭代的对象（例如 `array`、`map`、`set`、`string` 等）创建一个循环，并且有一个突出的优点，即优秀的可读性。

```
const arr = [3, 5, 7];
const str = 'hello';
for (let i of arr) {
   console.log(i); // 输出 3, 5, 7
}
for (let i of str) {
   console.log(i); // 输出 'h', 'e', 'l', 'l', 'o'
}
```

> 需要注意的是，请不要再生成器中使用 `for` `of`，因为即便循环体结束了，在结束一次循环以后，生成器会被关闭，并且在下一次循环中再次重复生成。

Author: ```suggestion
需要注意的是，请不要再生成器中使用 `for……of`，因为在一次循环的结束以后的下一次循环中，生成器会被再度调用，重复生成现有的结果。
```

> 需要注意的是，请不要再生成器中使用 `for` `of`，因为即便循环体结束了，在结束一次循环以后，生成器会被关闭，并且在下一次循环中再次重复生成。

Reviewer: > 

我不是很理解为什么在纠正错误之后，你还要把错误再改回来……你对这段话的理解存在问题。
1. 作者的意思是，不要在生成器中**再次**使用 `for…of`，或者说不要对生成器**重复**使用 `for…of`，而不是说在生成器中使用 `for…of`。
2. “因为在一次循环的结束以后的下一次循环中，生成器会被再度调用，重复生成现有的结果。”这句话完全是错误的，exiting the loop 表示已经退出了整个循环了，没有所谓的下一次循环这种说法。至于生成器被再度调用也是不存在的事情，在第一次退出循环之后生成器就已经关闭了（turn off），再对它使用for of 也是没有用的，不会产生新的值，更不可能“重复生成现有的结果”。

作者这一段也是摘抄的 MDN 的内容，所以具体的内容就阅读 MDN 的代码和解释吧，https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of，**Do not reuse generators** 这一小节。

> 需要注意的是，请不要再生成器中使用 `for` `of`，因为即便循环体结束了，在结束一次循环以后，生成器会被关闭，并且在下一次循环中再次重复生成。

Author: 修了，一开始没有代码，无法理解意思（，就以为是for of循环体里面重复调用generator生成一样的内容

### **4. `for` `in`**

`for` `in` 会在对象的所有可枚举属性上迭代指定的变量。对于每个不同的属性，`for` `in` 语句除返回数字索引外，还将返回用户定义的属性的名称。 因此，在遍历数组时最好使用带有数字索引的传统 `for` 循环。 因为 `for` `in` 语句还会迭代除数组元素之外的用户定义属性，就算我们修改了数组对象（例如添加自定义属性或方法），依然如此。

```
const details = {firstName: 'john', lastName: 'Doe'};
let fullName = '';
for (let i in details) {
    fullName += details[i] + ' '; // fullName: john doe
}
```

### `for` `of` 和 `for` `in`

`for` `of` 和 `for` `in` 之间的主要区别是它们迭代的内容。`for` `in` 循环遍历对象的属性，而 `for` `of` 循环遍历可迭代对象的值。

```
let arr= [4, 5, 6];
for (let i in arr) {
   console.log(i); // '0', '1', '2'
}
for (let i of arr) {
   console.log(i); // '4', '5', '6'
}
```

[由 [Tine Ivanič](https://unsplash.com/@tine999?utm_source=medium&utm_medium=referral) 上传至 [Unsplash](https://unsplash.com/?utm_source=medium&utm_medium=referral)](https://miro.medium.com/max/12000/0*E9FPH2LFeFnTGWF5)

## **Conclusion**

> - `for` 最快，但在可读性上比较差

Author: ```suggestion
- `for` 最快，但可读性比较差
```
